### PR TITLE
Request helper functions

### LINKED ISSUE #34: cover alexa (base) with tests
cover `/pkg/alexa/*.go` with tests (90+%)
* create useful tests (no need to test every single thing individually in every detail)
* review what's missing (compare to https://developer.amazon.com/de/docs/smapi/skill-manifest.html and https://developer.amazon.com/de/docs/custom-skills/request-and-response-json-reference.html)
** create a markdown file `alexa/TODO.md` which lists the missing elements/features

see #35

### PR DESCRIPTION
**Pull request by DrPsychick**
* use `golangci-lint` and fix all errors
* refactoring & cleanup
* update dependencies
* switch to go1.15+
* closes #34

### CODE REVIEW

File: pkg/alexa/request.go [hunk 1]
```diff
@@ -1,42 +1,60 @@
...
+var (
+	ErrorUnknown                   = errors.New("unknown error")
+	ErrorNoIntent                  = errors.New("request has no intent")
+	ErrorNoSlot                    = errors.New("slot does not exist")
+	ErrorSlotNoResolutions         = errors.New("slot has no resolutions")
+	ErrorSlotNoResolutionWithMatch = errors.New("no resolution with match")
+	ErrorNoSystemInContext         = errors.New("no system in context")
+	ErrorNoUserInContext           = errors.New("no user in context")
+	ErrorNoPersonInContext         = errors.New("no person in system context")
+	ErrorNoApplicationID           = errors.New("no application ID in the request")
```
Comment (nrwiersma) [inline pkg/alexa/request.go:15]: These should be named `ErrXxxx`.

File: pkg/alexa/request.go [hunk 2]
```diff
@@ -88,16 +135,82 @@ type SlotValue struct {
...
+// Slot returns the named slot or an error
+func (r *RequestEnvelope) Slot(name string) (Slot, error) {
+	i, err := r.Intent()
+	if err != nil {
+		return Slot{}, err
+	}
+
+	s, ok := i.Slots[name]
+	if !ok {
+		return Slot{}, ErrorNoSlot
```
Comment (nrwiersma) [inline pkg/alexa/request.go:161]: Perhaps a more general `ErrNotFound` would be easier to deal with for all of the cases you added. If you want to record what was not found in the different cases, you can make an error type `NotFoundError` that has a string for the type, like `slot`.

File: pkg/alexa/request.go [hunk 3]
```diff
@@ -141,16 +254,45 @@ type RequestType string
 
 // Request type constants.
 const (
-	// TypeLaunchRequest
+	// TypeLaunchRequest string that represents a launch request
```
Comment (nrwiersma) [inline pkg/alexa/request.go:257]:
```suggestion
	// TypeLaunchRequest defines a launch request type.
```

Apply below as well.

File: pkg/alexa/request.go [hunk 4]
```diff
@@ -222,12 +392,39 @@ type ContextSystem struct {
...
+// System returns the system object if it exists in the context
+func (r *RequestEnvelope) System() (*ContextSystem, error) {
+	if r.Context == nil || r.Context.System == nil {
+		return &ContextSystem{}, ErrorNoSystemInContext
+	}
+
+	return r.Context.System, nil
+}
+
+// ContextPerson returns the person in the context or throw an error if no person exists
```
Comment (nrwiersma) [inline pkg/alexa/request.go:404]: Go doesnt throw errors, it returns them.

File: pkg/alexa/request.go [hunk 5]
```diff
@@ -222,12 +392,39 @@ type ContextSystem struct {
...
+func (r *RequestEnvelope) ContextPerson() (*ContextSystemPerson, error) {
+	s, err := r.System()
+	if err != nil || s.Person == nil {
+		return &ContextSystemPerson{}, ErrorNoPersonInContext
+	}
+	return r.Context.System.Person, nil
+}
+
+// ContextUser returns the user in the context or throw an error if no user exists
```
Comment (nrwiersma) [inline pkg/alexa/request.go:413]: Same as above.

**Review by nrwiersma — CHANGES_REQUESTED** [review]

File: pkg/alexa/request.go (hunk 2 above)
Comment (DrPsychick) [reply pkg/alexa/request.go]: still working on it 😀 the golangci-lint still reports some errors that I have to take care of...

File: .travis.yml [hunk 6]
```diff
@@ -29,7 +28,10 @@ before_install:
   - docker pull lambci/lambda:go1.x
 
 script:
-  - golint ./...
+  # use golangci-lint
+  - GOLANGCI_LINTER_VERSION=v1.42.0
```
Comment (nrwiersma) [inline .travis.yml:32]: On Travis this should be in `env` and the line below should be in `before:`

File: cmd/alfalfa/make.go [hunk 7]
```diff
@@ -45,18 +46,22 @@ func runMake(c *cli.Context) error {
 			log.Fatal(ctx, err)
 		}
 		res, _ := json.MarshalIndent(s, "", "  ")
-		if err := ioutil.WriteFile("./alexa/skill.json", res, 0644); err != nil {
+		if err := ioutil.WriteFile("./alexa/skill.json", res, 0o644); err != nil { //nolint:gosec
```
Comment (nrwiersma) [inline cmd/alfalfa/make.go:49]: to get rid of the gosec you need to wrap the path in `filepath.Clean` if I remember correctly.

File: pkg/alexa/request.go [hunk 8]
```diff
@@ -314,10 +540,10 @@ type ContextViewportType struct {
 
 // Context represents the Alexa skill request context.
 type Context struct {
-	System      *ContextSystem         `json:"System,omitempty"`
+	System      *ContextSystem         `json:"System,omitempty"` //nolint:tagliatelle
```
Comment (nrwiersma) [inline pkg/alexa/request.go:543]: You should just disable this linter.

**Review by nrwiersma — CHANGES_REQUESTED** [review]

File: cmd/alfalfa/make.go (hunk 7 above)
Comment (DrPsychick) [reply cmd/alfalfa/make.go]: that didn't solve it. I needed to disable this check in the linter.

**Pull request merged**